sky/serve/serve_state.py

```diff
@@ -175,6 +178,11 @@ def from_replica_statuses(
         if sum(status2num[status]
                for status in ReplicaStatus.failed_statuses()) > 0:
             return cls.FAILED
+        # When min_replicas = 0, there is no (provisioning) replica.
+        if (len(replica_statuses) - sum(
+            [status2num[status] for status in ReplicaStatus.failed_statuses()
+            ]) == 0):
```

Comment:
Also, what if there are some replicas being preempted or not ready? Feeling like there are too many edge cases. Should we just return REPLICA_INIT if we find any PENDING/PROVISIONING/STARTING replicas and otherwise return NO_REPLICAS?

Reply:
I think the current if-statements should take care of replicas being not ready, and the two ways to write it seem to be the same?

Reply:
Actually, this still seems to be better
```
    @classmethod
    def from_replica_statuses(
            cls, replica_statuses: List[ReplicaStatus]) -> 'ServiceStatus':
        status2num = collections.Counter(replica_statuses)
        # If one replica is READY, the service is READY.
        if status2num[ReplicaStatus.READY] > 0:
            return cls.READY
        if sum(status2num[status]
               for status in ReplicaStatus.failed_statuses()) > 0:
            return cls.FAILED
        # When min_replicas = 0, there is no (provisioning) replica.
        if len(replica_statuses) == 0:
            return cls.NO_REPLICA
        return cls.REPLICA_INIT
```
When there is only failed replicas, it should return FAILED instead of NO_REPLICA, even when only failed replicas are left.

Reply:
Oh that is true. SG!